pkg/controller/appdefinition/pullappimage.go

```diff
@@ -96,3 +112,62 @@ func determineTargetImage(appInstance *v1.AppInstance) (string, string) {
 		}
 	}
 }
+
+const (
+	AppImageResolutionFailureEventType = "AppImageResolutionFailure"
+	AppImageResolutionSuccessEventType = "AppImageResolutionSuccess"
+)
+
+// AppImageResolutionEventDetails captures additional info about App image resolution.
+type AppImageResolutionEventDetails struct {
+	// AppResourceVersion is the resourceVersion of the App the image is being resolved for.
+	AppResourceVersion string `json:"appResourceVersion"`
+
+	// TargetImage is the image being resolved.
+	TargetImage string `json:"targetImage,omitempty"`
+
+	// ResolvedImage is the image, post resolution.
+	// +optional
+	ResolvedImage string `json:"resolvedImage,omitempty"`
+
+	// Err the error that occurred during resolution, if any.
+	// +optional
+	Err string `json:"err,omitempty"`
+}
+
+func recordResolutionEvent(ctx context.Context, recorder event.Recorder, obj kclient.Object, err error, targetImage, resolvedImage string) {
```

Comment:
Shared directly, but wanted to post here too. this definitely makes sense.

Here's some additional information we could add:
- Whether this was triggered by auto-upgrade or not (you could make `determineTargetImage` return an extra field to accomplish this)
- app.status.appImage.Digest and app.status.appImage.VCS
- Would it be overkill to also record the "previous values" for targetImage, resolvedImage, digest, and vcs? that would be fairly powerful to have
- TargetImage and ResolvedImage suck as names to expose to the user, but i think the meaning of these two fields is slightly different depending on whether the app auto-upgrades or not. let me confirm that and get back to you with slightly better names

Reply:
Ok, ran through a few scenarios. "ResolvedImage" is kind of an implementation detail that I think we should drop. It is used to get an image ID if the image in question is "local" (not from a remote registry). If the image ISNT local, the value is the same as targetImage. Furthermore, the ID happens to be the same value as the digest field with the "sha256:" prefix stripped off. So, I dont think it holds much value. Just drop it.


I think targetImage is a good name to keep for the other field, or could just call it image.

Reply:
I refactored this and tried to hit all your suggestions.

The largest semantic change is moving the "topic" of the events generated here from "Resolution" to "Pull".